pallas-primitives/src/lib.rs

```diff
@@ -1,10 +1,392 @@
+pub struct ExUnits {
+    #[n(0)]
+    pub mem: u64,
+    #[n(1)]
+    pub steps: u64,
+}
+
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub struct ExUnitPrices {
+    #[n(0)]
+    pub mem_price: PositiveInterval,
+
+    #[n(1)]
+    pub step_price: PositiveInterval,
+}
+
+pub type Genesishash = Bytes;
+
+pub type GenesisDelegateHash = Bytes;
+
+pub type IPv4 = Bytes;
+
+pub type IPv6 = Bytes;
+
+pub type Metadata = KeyValuePairs<MetadatumLabel, Metadatum>;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum Metadatum {
+    Int(Int),
+    Bytes(Bytes),
+    Text(String),
+    Array(Vec<Metadatum>),
+    Map(KeyValuePairs<Metadatum, Metadatum>),
+}
+
+impl<'b, C> minicbor::Decode<'b, C> for Metadatum {
+    fn decode(d: &mut minicbor::Decoder<'b>, ctx: &mut C) -> Result<Self, minicbor::decode::Error> {
+        match d.datatype()? {
+            minicbor::data::Type::U8
+            | minicbor::data::Type::U16
+            | minicbor::data::Type::U32
+            | minicbor::data::Type::U64
+            | minicbor::data::Type::I8
+            | minicbor::data::Type::I16
+            | minicbor::data::Type::I32
+            | minicbor::data::Type::I64
+            | minicbor::data::Type::Int => {
+                let i = d.decode()?;
+                Ok(Metadatum::Int(i))
+            }
+            minicbor::data::Type::Bytes => Ok(Metadatum::Bytes(d.decode_with(ctx)?)),
+            minicbor::data::Type::String => Ok(Metadatum::Text(d.decode_with(ctx)?)),
+            minicbor::data::Type::Array | minicbor::data::Type::ArrayIndef => {
+                Ok(Metadatum::Array(d.decode_with(ctx)?))
+            }
+            minicbor::data::Type::Map | minicbor::data::Type::MapIndef => {
+                Ok(Metadatum::Map(d.decode_with(ctx)?))
+            }
+            _ => Err(minicbor::decode::Error::message(
+                "Can't turn data type into metadatum",
+            )),
+        }
+    }
+}
+
+impl<C> minicbor::Encode<C> for Metadatum {
+    fn encode<W: minicbor::encode::Write>(
+        &self,
+        e: &mut minicbor::Encoder<W>,
+        ctx: &mut C,
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        match self {
+            Metadatum::Int(a) => {
+                e.encode_with(a, ctx)?;
+            }
+            Metadatum::Bytes(a) => {
+                e.encode_with(a, ctx)?;
+            }
+            Metadatum::Text(a) => {
+                e.encode_with(a, ctx)?;
+            }
+            Metadatum::Array(a) => {
+                e.encode_with(a, ctx)?;
+            }
+            Metadatum::Map(a) => {
+                e.encode_with(a, ctx)?;
+            }
+        };
+
+        Ok(())
+    }
+}
+
+pub type MetadatumLabel = u64;
+
+#[derive(
+    Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy,
+)]
+#[cbor(index_only)]
+pub enum NetworkId {
```

Comment:
This is highly cardano-specific. APEX Fusion networks use the following

```
prime testnet: networkId == 0
prime mainnet: networkId == 1
vector testnet: networkId == 2
vector mainnet: networkId == 3
```

Does this code limit us to using only these specific values with pallas?

Reply:
Not sure to see the relevance of APEX here? I mean, 90% of the content of pallas-primitives is Cardano-specific 😅 ...

Reply:
If we kept this as Alonzo-specific, I would follow Alonzo CDDL and constrains this to 0 / 1.

Since we're moving this a more generic primitive, I would turn this into an int and leave the semantics of each value to a higher layer.